aeneas/src/ir/VariantSolver.v3

```diff
@@ -190,68 +217,138 @@ class VariantSolver(nc: NormalizerConfig, rn: ReachabilityNormalizer, verbose: b
 		var ta = vecT.extract();
 		var vn = VariantNorm.new(rc.oldType, Tuple.newType(Lists.fromArray(ta)), ta, fields, null);
 		vn.fieldRanges = fieldRanges;
+		vn.vecO = vecO.extract();
 		rc.variantNorm = vn;
 
 		if (verbose) Terminal.put1("variant norm %q\n", rc.variantNorm.render);
 	}
+	// Perform recursive backtracking on the potential representation.
+	// With each scalar in each case, we have several choices to make:
+	// 1. Merge this scalar with an existing scalar in the representation that hasn't been used by the current case.
+	// 2. Append this scalar as a new scalar.
+	// We can only determine distinguishability after all the scalars have been assigned. Heuristics will help speed this part up.
+	var assignments: HashMap<CaseField, int>;
+	private def tryRepresentationForField(curRep: Vector<Scalar.set>, curUsed: Vector<bool>, fields: Array<Array<Type>>, curCase: int, curField: int) -> bool {	
+		if (curCase >= fields.length) {
+			// XXX: check distinguishable
+			return true; // true, since we add a tag scalar - this will change once packing is involved
+		}
+		if (curField >= fields[curCase].length) {
+			// move on to next case
+			curUsed = Vector<bool>.new();
+			for (i < curRep.length) curUsed.put(false);
+			return tryRepresentationForField(curRep, curUsed, fields, curCase + 1, 0);
+		}
+		
+		var sc = nc.GetScalar(rn.ra.compiler, rn.ra.prog, fields[curCase][curField]);
+		var none: Scalar.set;
+
+		for (i < curRep.length) {
+			if (curUsed[i]) continue;
+			var oldRep = curRep[i];
+
+			if ((sc & oldRep) != none) {
+				curUsed[i] = true; curRep[i] = sc & oldRep;
+				assignments[CaseField(curCase, curField)] = i;
+				if (tryRepresentationForField(curRep, curUsed, fields, curCase, curField + 1)) return true;
+				curUsed[i] = false; curRep[i] = oldRep;
+			}
+		}
+
+		var len = curRep.length;
+		curRep.put(sc); curUsed.put(true);
+		assignments[CaseField(curCase, curField)] = curRep.length - 1;
+		if (tryRepresentationForField(curRep, curUsed, fields, curCase, curField + 1)) return true;
+		curRep.resize(len); curUsed.resize(len);
+
+		return false;
+	}
 	private def unboxUsingTaggedVariantNorm(rc: RaClass) -> bool {
 		var vecT = Vector<Type>.new();
 		var parentUnboxed = rc.orig.boxing == Boxing.UNBOXED;
 
-		var scalars = ScalarSet.new(nc.UnifyRepresentations(rn.ra.compiler, rn.ra.prog, _, _));
-		var cases: List<(Array<VariantField>, Array<(int, int)>)>;
+		var numChildren = Lists.length(rc.children);
 
+		var cases = Array<(Array<VariantField>, Array<(int, int)>, Array<(int, int)>, Array<Type>)>.new(numChildren);
```

Comment:
The `cases` array holds information for each child and is indexed in parallel to the children list. Does it make sense to instead put some state into `RaClass` that is only used here? It would simplify the iteration over children and the parallel array.